### PR TITLE
Snapshot support.  API:

### PR DESCRIPTION
**Pull request by argv0**
e_leveldb:snapshot(db_ref()) -> {ok, snapshot_ref()}.
e_leveldb:snapshot_close(snapshot_ref()) -> ok.
e_leveldb:get(db_ref()|snapshot_ref(), ...) -> ...
e_leveldb:iterator(db_ref()|snapshot_ref(), ..) -> ...
e_leveldb:write(..., [{return_snapshot, true}]) -> {ok, snapshot_ref()}.
e_leveldb:write(snapshot_ref(), ...) -> {error, read_only_snapshot}.

Also added support for e_leveldb:open() options:

{paranoid_checks, boolean()}
{max_open_files, pos_integer()}
{write_buffer_size, pos_integer()}
{block_size, pos_integer()}
{block_restart_interval, pos_integer()}
{cache_size, pos_integer()}

### CODE REVIEW

File: c_src/e_leveldb.cc [hunk 1]
```diff
@@ -192,6 +239,43 @@ ERL_NIF_TERM error_tuple(ErlNifEnv* env, ERL_NIF_TERM error, leveldb::Status& st
                             enif_make_tuple2(env, error, reason));
 }
 
+
+/* make_snapshot_handle retains the db_handle argument, and does no ownership
+   transfer of the handle it returns, so it must be released before returning
+   to erlang 
+*/
+e_leveldb_snapshot_handle* make_snapshot_handle(e_leveldb_db_handle* db_handle,
```
Comment (dizzyd) [inline c_src/e_leveldb.cc]: I believe this function should be marked as static to indicate it's an internal function.

File: c_src/e_leveldb.cc [hunk 2]
```diff
@@ -192,6 +239,43 @@ ERL_NIF_TERM error_tuple(ErlNifEnv* env, ERL_NIF_TERM error, leveldb::Status& st
...
+}
+
+int extract_handles(ErlNifEnv *env, ERL_NIF_TERM term, e_leveldb_db_handle **db_handle,
+                    e_leveldb_snapshot_handle **snapshot_handle) { 
+    *db_handle = 0;
+    *snapshot_handle = 0;
+    if (enif_get_resource(env, term, e_leveldb_db_RESOURCE, (void**)db_handle))
+        return 1;
+    else if (enif_get_resource(env, term, e_leveldb_snapshot_RESOURCE, 
+                               (void **)snapshot_handle)) {
```
Comment (dizzyd) [inline c_src/e_leveldb.cc]: For consistency, the brackets need to be on their own line.

**Comment by matthewvon** [comment]
Is this pull request still valid / needed?  Or should I close it?

**Comment by benoitc** [comment]
is there any work in progress to add snapshot support? Was this PR closed for any reason except no answer from @argv0 ?

**Comment by matthewvon** [comment]
No progress.   Basho has no interest in the snapshot support.  Snapshots via the iterators was the better solution for us.

**Comment by benoitc** [comment]
would you accept a patch for it?

**Comment by matthewvon** [comment]
Why?  This is an open source branch.  You can fork to a personal repository and modify to your heart's content.

**Comment by benoitc** [comment]
odd enough but ok. thanks!

**Comment by benoitc** [comment]
just to clarify my comment, i was expecting sending a pr here as the canonical repositoy and wanted to make sure it would be discussed since all the work around that have been closed. Now it seems this is not.

**Comment by matthewvon** [comment]
Sending a PR to basho/eleveldb or to basho/leveldb is requesting that we (namely I) take ownership of the feature in perpetuity.  We have done that in a few cases.  One case for leveldb about this time last year resulted in a huge bug (significant performance loss under heavy load) because the merge (by me) was not clean.  

Also when we have taken public PRs, I have required a full 6 day test cycle on each.  There is no time on the horizon for me to give a PR that kind of attention, especially when it does not contribute to our paying product Riak.  The upcoming Time Series release has the entire backends team swamped.

This is NOT to say your request is without merit.  Just giving you the cost / benefit tradeoffs that exist at this time.

Matthew

**Comment by benoitc** [comment]
@matthewvon ok fair enough :) will see how it can go further, thanks for your response.

**Pull request closed without merging**